components/tidb_query_executors/src/index_lookup_executor.rs

```diff
@@ -0,0 +1,2287 @@
+    table_task_iter_builder: Option<Builder>,
+    table_scan_params: TableScanParams,
+
+    phase: IndexLookUpPhase<Builder::Iterator, F>,
+    intermediate_results: Vec<BatchExecuteResult>,
+}
+
+impl<S, Src, Builder, F> BatchIndexLookUpExecutor<S, Src, Builder, F>
+where
+    S: Storage,
+    Src: BatchExecutor<StorageStats = S::Statistics>,
+    Builder: TableTaskIterBuilder<Iterator: TableTaskIterator<Storage = S>>,
+    F: KvFormat,
+{
+    const DEFAULT_INDEX_LOOKUP_BATCH_SIZE: usize = 256;
+
+    // TODO: remove #[allow(dead_code)] after new function is used
+    #[allow(dead_code)]
+    #[inline]
+    pub fn new(
+        config: Arc<EvalConfig>,
+        src: Src,
+        table_scan_params: TableScanParams,
+        table_task_iter_builder: Option<Builder>,
+    ) -> Self {
+        let force_no_index_lookup =
+            if config.paging_size.is_some() || table_task_iter_builder.is_none() {
+                // We did not support index lookup when paging is enabled.
+                // TODO: support paging
+                // some times we do not have table_task_iter_builder, such as
+                // - CommonHandle
+                // TODO: support CommonHandle
+                true
+            } else {
+                false
+            };
+
+        let output_schema = table_scan_params
+            .columns_info
+            .iter()
+            .map(field_type_from_column_info)
+            .collect::<Vec<_>>();
+
+        BatchIndexLookUpExecutor {
+            output_schema,
+            config,
+            force_no_index_lookup,
+            table_lookup_batch_size: Self::DEFAULT_INDEX_LOOKUP_BATCH_SIZE,
+            src,
+            src_is_drained: BatchExecIsDrain::Remain,
+            intermediate_results: vec![],
+            table_task_iter_builder,
+            phase: IndexLookUpPhase::default(),
+            table_scan_params,
+        }
+    }
+
+    #[inline]
+    fn empty_result(is_drained: BatchExecIsDrain) -> BatchExecuteResult {
+        BatchExecuteResult {
+            physical_columns: LazyBatchColumnVec::empty(),
+            logical_rows: vec![],
+            warnings: EvalWarnings::default(),
+            is_drained: Ok(is_drained),
+        }
+    }
+
+    #[inline]
+    fn err_result(err: Error) -> BatchExecuteResult {
+        BatchExecuteResult {
+            physical_columns: LazyBatchColumnVec::empty(),
+            logical_rows: vec![],
+            warnings: EvalWarnings::default(),
+            is_drained: Err(err),
+        }
+    }
+
+    #[inline]
+    fn step_to_table_lookup(&mut self, warnings: &mut EvalWarnings) -> Result<()> {
+        let state = self.phase.mut_index_scan_or_err()?;
+
+        let builder = self.table_task_iter_builder.as_ref().map_or_else(
+            || {
+                Err(other_err!(
+                    "No table task builder available for index lookup"
+                ))
+            },
+            |p| Ok(p),
+        )?;
```

Comment:
I think `ok_or` is what you're looking for.

```suggestion
        let builder = self.table_task_iter_builder.as_ref().ok_or(Err(other_err!(
            "No table task builder available for index lookup"
        )))?;
```